fx_compile.py

```diff
@@ -0,0 +1,432 @@
+            new_idxs = idxs[:]
+            new_idxs[dim] -= to_expr(sm)
+            load = te.ifThenElse(idxs[dim] < to_expr(sm), load, tensor.load(new_idxs))
+        return load
+    return te.Compute(name, get_dim_args(out_shape), f)
+
+lowering_functions[torch.transpose] = transpose_lower
+lowering_functions[torch.flatten] = flatten_lower
+lowering_functions[torch.cat] = cat_lower
+
+def bmm_lower(name, out_shape, inp_shapes, args):
+    M1 = args[0]
+    M2 = args[1]
+    B, N, M = inp_shapes[0][0]
+    P = inp_shapes[1][0][2]
+
+    def f(b, n, p, m):
+        return M1.load([b, n, m]) * M2.load([b, m, p])
+    mm = te.Compute('mm', get_dim_args([B,N,P,M]), f)
+    return te.SumReduce(name, get_dim_args([B, N, P]), mm, get_dim_args([M]))
+
+
+def mm_lower(name, out_shape, inp_shapes, args):
+    M1 = args[0]
+    M2 = args[1]
+    N, M = inp_shapes[0][0]
+    P = inp_shapes[1][0][1]
+
+    def f(n, p, m):
+        return M1.load([n, m]) * M2.load([m, p])
+    mm = te.Compute('mm', get_dim_args([N,P,M]), f)
+    return te.SumReduce(name, get_dim_args([N, P]), mm, get_dim_args([M]))
+
+lowering_functions[torch.bmm] = bmm_lower
+lowering_functions[torch.mm] = mm_lower
+
+
+
+def lower_function(name, op, shape_env, nnc_args, args):
+    inp_shapes = fx.node.map_aggregate(args, lambda arg: shape_env[arg.name] if isinstance(arg, fx.Node) else None)
+    return lowering_functions[op](name, shape_env[name][0], inp_shapes, nnc_args)
+
+def nnc_compile(model: torch.nn.Module, shape_specialize, shape_env) -> torch.nn.Module:
+    fx_model = fx.symbolic_trace(model)
+    env = {}
+    def get_te_shapes(name):
+        return [te.ExprHandle.int(i) for i in shape_env[name][0]]
+
+    def get_nnc_type(dtype):
+        if dtype == torch.float:
+            return torch._C.te.Dtype.Float
+        elif dtype == torch.long:
+            return torch._C.te.Dtype.Long
+        else:
+            raise RuntimeError("nyi")
+
+    def get_te_type(name):
+        return get_nnc_type(shape_env[name][1])
+
+    def gen_compute(args):
+        te_args = [env[arg.name] for arg in args]
+
+    def lookup_env(l):
+        return fx.node.map_aggregate(l, lambda x: env[x.name] if isinstance(x, fx.Node) else x)
+
+    def fetch_attr(target : str):
+        target_atoms = target.split('.')
+        attr_itr = fx_model
+        for i, atom in enumerate(target_atoms):
+            if not hasattr(attr_itr, atom):
+                raise RuntimeError(f"Node referenced nonexistant target {'.'.join(target_atoms[:i])}")
+            attr_itr = getattr(attr_itr, atom)
+        return attr_itr
+
+    outs = None
+    inputs = []
+    module_attrs = []
+    for node in fx_model.graph.nodes:
+        if node.op == 'placeholder':
+            shapes = get_te_shapes(node.name)
+            env[node.name] = te.Placeholder(node.name, get_te_type(node.name), shapes)
+            inputs.append(env[node.name])
+        elif node.op == 'call_function':
```

Comment:
I understand this is a prototype, but this also brings up the question of supporting methods (probably not too hard, but ideally we'd normalize those into functional forms) and modules (would we have to make all the modules functional and trace down into them? or have a canonical TE lowering for all of them, or something else)

Reply:
From "me thinking about it right now", I don't think modules need to be functional. Decomposing it would certainly make it easier to handle all of them, but you could also write TE lowerings without too much issue.

From my understanding, making modules functional is primarily only an issue for `autograd`, as that's the pass where the API needs access to module internals.

The main reason a "module functionalization" pass is needed is since you need the gradient of the output with respect to the module's weights. So if you had 

```
layer = nn.Linear(5,5)
def forward(x):
      return layer(x)
```

calling `grad(forward)(x)` gets you the gradient of `x` (the input), not `layer.weight`. What you would really want is `grad(forward)(x, layer.weight)`, but that's not really possible right now for a variety of reasons.

For a lowering pass, or for vmap, there's no issues with just accessing the module attributes and decomposing it into

```
def forward(x):
    return x @ layer.weight + layer.bias
```

You can see that in the current DeepAndWide there's module attributes I'm accessing that are getting "compiled".

Reply:
My comment was more about maintainability in the codebase. The way I see it there are two options:

* Have a separate canonical TE lowering for each standard library `Module` instance. This is prohibitive and probably not a good idea, as each `Module` will need to be written twice and those implementations will need to stay in sync.
* Have the module written once and automatically lowerable to TE. I think there are sub-possibilities here:
  a. Make it so that the `Modules` are convertible to TE, such as via symbolic tracing + a lowering on the RISCy primitives. This would potentially require modification of the way we currently write `Modules` so that they would be captureable. Specifically, for them to be captureable in FX they would need to be functional (in the sense of the ops are functional, not the sense that they don't have attributes).
  b. Make it so that the `Module`s are written in TE in the first place. This would be a huge change and potentially not in the spirit of being Pythonic, but maybe things will go fast.

Reply:
Ah, I see what you mean. Yes, ideally we would make it so that `Modules` are convertible to TE. I think your work showed that this isn't actually that bad, especially if we're shape specializing.

There might be a couple we need special handling for, but overall it seems acceptable.